addons/l10n_multilang/models/l10n_multilang.py

```diff
@@ -88,20 +89,28 @@ def _process_fiscal_pos_translations(self, company_id, langs, field):
 
     def _get_template_from_model(self, company_id, model):
         """ Find the records and their matching template """
-        # genereated records have an external id with the format <company id>_<template id>
-        out_data = self.env['ir.model.data'].search([('model', '=', model), ('name', '=like', str(company_id)+'\_%')])
-        out_ids = self.env[model].browse(out_data.mapped('res_id'))
-
-        # templates and records may have been created in a different order
-        # reorder them based on external id names
-        in_xml_id_names = [xml_id.partition(str(company_id) + '_')[-1] for xml_id in out_data.mapped('name')]
-        in_xml_id_names = {name: index for index, name in enumerate(in_xml_id_names)}
-
-        in_xml_ids = self.env['ir.model.data'].search([('model', '=', model+'.template'), ('name', 'in', list(in_xml_id_names))])
-        in_xml_ids = in_xml_ids.sorted(key=lambda x: in_xml_id_names[x.name])
-
-        in_ids = self.env[model+'.template'].browse(in_xml_ids.mapped('res_id'))
-        return (in_ids, out_ids)
+        # generated records have an external id with the format <company id>_<template xml id>
+        grouped_out_data = defaultdict(lambda: self.env['ir.model.data'])
+        for imd in self.env['ir.model.data'].search([('model', '=', model),
+                                                      ('name', '=like', str(company_id)+'\_%')]):
+            grouped_out_data[imd.module] += imd
+
+        in_records = self.env[model+'.template']
+        out_records = self.env[model]
+        for module, out_data in grouped_out_data.items():
+            # templates and records may have been created in a different order
+            # reorder them based on external id names
+            expected_in_xml_id_names = {xml_id.name.partition(str(company_id) + '_')[-1]: xml_id for xml_id in out_data}
+
+            in_xml_ids = self.env['ir.model.data'].search([('model', '=', model+'.template'),
+                ('module', '=', module), ('name', 'in', list(expected_in_xml_id_names))])
+            in_xml_ids = {xml_id.name: xml_id for xml_id in in_xml_ids}
+
+            for name in expected_in_xml_id_names:
+                in_records += self.env[model+'.template'].browse( in_xml_ids[name].res_id )
+                out_records += self.env[model].browse( expected_in_xml_id_names[name].res_id )
```

Comment:
iterate on `expected_in_xml_id_names.items()` and save a lookup for each out_record?